rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+    If `replication_mode` == `PER_WORKER`, this is the worker ID.
+    If `replication_mode` == `PER_REPLICA`, this is the compute replica ID.
+    """
+
+  @property
+  def num_input_pipelines(self) -> int:
+    """Returns the number of input pipelines.
+
+    If `replication_mode` == `SINGLE`, this is always `1`.
+    If `replication_mode` == `PER_WORKER`, this is the number of workers.
+    If `replication_mode` == `PER_REPLICA`, this is `num_replicas`.
+    """
+```
+
+#### InputIterator
+
+An instance of `InputIterator` is returned by `prepare_input` and can be passed
+to `run`.
+
+```python
+class InputIterator(Generic[T]):
+  """An input iterator, intended to be passed to `Replicator.run`."""
+
+  def get_next(self) -> List[T]:
+    """Returns the next inputs for each replica."""
+
+  def reinitialize(self) -> List[tf.Operation]:
+    """Re-initializes the inputs."""
+```
+
+### Model Parallelism
+
+Model parallelism can allow the execution of models when there is insufficient
+memory on a single device (and further data parallelism is not possible).
+
+Replicator supports model partitioning, a form of model parallelism in which the
+ops within the model are placed onto multiple devices. This is done via the
+`logical_device` context.
+
+Inputs are always located on logical device 0. Ops within the step function
+default to logical device 0 (except on TPU, which defaults to automatic
+placement).
+
+```python
+with replicator.scope():
+  with replicator.logical_device_for_variables(1):
+    v = tf.get_variable(...)  # On logical device 1.
+    x = v * 3  # Ignores logical device context.
+
+def step(ctx, inputs):  # `inputs` on logical device 0.
+  a = model_part1(inputs)  # Implicitly on logical device 0.
+  with ctx.logical_device(0):
+    b = model_part2(a)  # Explicitly on logical device 0.
+  with ctx.logical_device(1):
+    return model_part3(b)  # Explicitly on logical device 1.
+```
+
+### Eager mode support
+
+Eager mode support mostly drops out of building atop of `DistributionStrategy`.
+The entire API is designed to be eager compatible. We've called out the
+difference in behavior in eager and graph mode in the specific API
+documentations. In general, in graph mode, many of the APIs return ops which
+should be later executed with `session.run`, and in eager mode, they simply run
+the ops.
+
+## Usage Examples
+
+### Building a Replicator object
+
+The first step is to build a `Replicator` instance for the chosen platform:
+
+```python
+dist_strat = ...  # Create `DistributionStrategy` for a specific regime.
+replicator = tf.replicator.Replicator(dist_strat)
+```
+
+### Simple Classification
+
+Below is a simple usage example for an image classification use case.
+
+#### Training
+
+```python
+with replicator.scope():
+  model = resnet.ResNetV1(resnet.BLOCKS_50)
+  optimizer = tf.train.MomentumOptimizer(learning_rate, 0.9)
+
+def input_fn(ctx):
+  assert effective_batch_size % ctx.num_replicas_in_sync == 0
```

Comment:
In practice, if this assert fails, what happens? Or, more broadly, how does error propagation work? Would this stop all replicas, or if one fails, the rest continue?

Reply:
since this type of error will happen during graph construction, in the in-graph case, this will just mean that the graph will not be constructed and no replicas will start training. (for multi-GPU and TPU strategy, this would be the case)
For the between-graph case (which is typically used for multi worker), there are a couple different modes - standalone client and independent worker mode. We don't talk about these in detail in this design, but there will be an upcoming design talking about how to setup the cluster etc for multi worker cases easily. 
@yuefengz - can you comment on the current state of error propagation for between graph cases for multi worker training?

Reply:
Yeah this case will fail on all of them. But Karmel's question is also about broad error propagation - if graph creation fails on one of the workers but not on all?

Reply:
There is not any existing mechanism to broadcast errors. That depends on the outer loop. E.g. Kubeflow can monitor these workers and kill the whole training cluster if any of the worker fails with non-retryable errors.